### Завдання 1

Проведіть класифікацію на основі датасету UCI ML Breast Cancer Wisconsin, використовуючи вказані у презентації моделі бустингу, беггінгу та стекінгу. Змініть їх параметри наступним чином:
у StackingClassifier як базові алгоритми встановіть LogisticRegression та DecisionTreeClassifier, як фінальну модель – SVC;
у BaggingClassifier як базовий алгоритм встановіть DecisionTreeClassifier;
у AdaBoostClassifier як базовий алгоритм встановіть LogisticRegression
у XGBСlassifier як базовий алгоритм (booster) встановіть ‘gbtree’.
У зведеній таблиці порівняйте результати класифікацій, наведених у коді, із класифікаціями, що отримані вами.


In [71]:
from sklearn.ensemble import StackingClassifier
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import AdaBoostClassifier
from xgboost import XGBClassifier
import pandas as pd

#### Підготуємо дані для аналізу моделей

In [72]:
data, target = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.3, random_state=12)

#### Проведемо аналіз даних за допомогою ансамблю StackingClassifier з базовими моделями LogisticRegression і SGDClassifier та фінальною моделью DecisionTreeClassifier.

In [73]:
estimators = [('Logistic_regrssion', LogisticRegression()), ('SGDClassifier', SGDClassifier())]
model_stc = StackingClassifier(estimators=estimators, final_estimator=DecisionTreeClassifier())
model_stc.fit(X_train, y_train)

C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check

StackingClassifier(estimators=[('Logistic_regrssion', LogisticRegression()),
                               ('SGDClassifier', SGDClassifier())],
                   final_estimator=DecisionTreeClassifier())

In [74]:
y_pred = model_stc.predict(X_test)
m11=model_stc.score(X_test, y_test)
print(m11)

0.9064327485380117


#### Тепер проведемо аналогічний аналіз даних за допомогою ансамблю StackingClassifier з базовими моделями LogisticRegression і DecisionTreeClassifier та фінальною моделью SVC.

In [75]:
estimators = [('Logistic_regrssion', LogisticRegression()), ('Decision_tree_classifier', DecisionTreeClassifier())]
model_stc = StackingClassifier(estimators=estimators, final_estimator=SVC())
model_stc.fit(X_train, y_train)

C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check

StackingClassifier(estimators=[('Logistic_regrssion', LogisticRegression()),
                               ('Decision_tree_classifier',
                                DecisionTreeClassifier())],
                   final_estimator=SVC())

In [76]:
y_pred = model_stc.predict(X_test)
m12=model_stc.score(X_test, y_test)
print(m12)

0.9415204678362573


#### Проведемо аналіз даних за допомогою BaggingClassifier з базовою моделью LogisticRegression.

In [77]:
model_bclf = BaggingClassifier(base_estimator=LogisticRegression(), n_estimators=20, random_state=12)
model_bclf.fit(X_train, y_train)

C:\Python\Python310\lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/pre

C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check

BaggingClassifier(base_estimator=LogisticRegression(), n_estimators=20,
                  random_state=12)

In [78]:
m21=model_bclf.score(X_test, y_test)
print(m21)

0.935672514619883


#### Тепер проведемо порівняння даних отриманих за допомогою BaggingClassifier з базовою моделью DecisionTreeClassifier.

In [79]:
model_bclf = BaggingClassifier(base_estimator=DecisionTreeClassifier(max_depth=6), n_estimators=20, random_state=12)
model_bclf.fit(X_train, y_train)

C:\Python\Python310\lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


BaggingClassifier(base_estimator=DecisionTreeClassifier(max_depth=6),
                  n_estimators=20, random_state=12)

In [80]:
m22=model_bclf.score(X_test, y_test)
print(m22)

0.9064327485380117


#### Проведемо аналіз даних за допомогою AdaBoostClassifier з базовою моделью DecisionTreeClassifier.

In [81]:
mole_ada_bc = AdaBoostClassifier(base_estimator=DecisionTreeClassifier(max_depth=6), n_estimators=100, random_state=12)
mole_ada_bc.fit(X_train, y_train)

C:\Python\Python310\lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


AdaBoostClassifier(base_estimator=DecisionTreeClassifier(max_depth=6),
                   n_estimators=100, random_state=12)

In [82]:
m31=mole_ada_bc.score(X_test, y_test)
print(m31)

0.9181286549707602


#### Порівняємо дані отримані за допомогою AdaBoostClassifier з базовою моделью LogisticRegression.

In [83]:
mole_ada_bc = AdaBoostClassifier(base_estimator=LogisticRegression(), n_estimators=100, random_state=12)
mole_ada_bc.fit(X_train, y_train)

C:\Python\Python310\lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/pre

AdaBoostClassifier(base_estimator=LogisticRegression(), n_estimators=100,
                   random_state=12)

In [84]:
m32=mole_ada_bc.score(X_test, y_test)
print(m32)

0.935672514619883


#### Проведемо аналіз даних за допомогою XGBClassifier з базовою моделью booster = 'gblinear'.

In [85]:
model_xgbcls = XGBClassifier(booster = 'gblinear', learning_rate=0.02, n_estimators=10)
model_xgbcls.fit(X_train, y_train)

XGBClassifier(base_score=None, booster='gblinear', callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.02, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=10, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=None, ...)

In [86]:
m41=model_xgbcls.score(X_test, y_test)
print(m41)

0.8654970760233918


#### Порівняємо дані отримані за допомогою XGBClassifier з базовою моделью booster = 'gbtree'.

In [87]:
model_xgbcls = XGBClassifier(booster = 'gbtree', learning_rate=0.02, n_estimators=10)
model_xgbcls.fit(X_train, y_train)

XGBClassifier(base_score=None, booster='gbtree', callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.02, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=10, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=None, ...)

In [88]:
m42=model_xgbcls.score(X_test, y_test)
print(m42)

0.9122807017543859


In [89]:
tab=[['LogisticRegression, SGDClassifier, final-DecisionTreeClassifier',m11,None,None,None],
     ['LogisticRegression,  DecisionTreeClassifier, final-CVS',m12,None,None,None],
     ['BaggingClassifier/LogisticRegression',None,m21,None,None],
     ['BaggingClassifier/DecisionTreeClassifier',None,m22,None,None],
     ['AdaBoostClassifier/DecisionTreeClassifier',None,None,m31,None],
     ['AdaBoostClassifier/LogisticRegression',None,None,m32,None],
     ['XGBClassifie/booster = gblinear',None,None,None,m41],
     ['XGBClassifie/booster = gbtree',None,None,None,m42],
]


In [103]:
table=pd.DataFrame(tab, columns=['Ansamble of model','StackingClassifier', 'BaggingClassifier', 'AdaBoostClassifier','XGBClassifier' ])
table

Ansamble of model  StackingClassifier  \
0  LogisticRegression, SGDClassifier, final-Decis...            0.906433   
1  LogisticRegression,  DecisionTreeClassifier, f...            0.941520   
2               BaggingClassifier/LogisticRegression                 NaN   
3           BaggingClassifier/DecisionTreeClassifier                 NaN   
4          AdaBoostClassifier/DecisionTreeClassifier                 NaN   
5              AdaBoostClassifier/LogisticRegression                 NaN   
6                    XGBClassifie/booster = gblinear                 NaN   
7                      XGBClassifie/booster = gbtree                 NaN   

   BaggingClassifier  AdaBoostClassifier  XGBClassifier  
0                NaN                 NaN            NaN  
1                NaN                 NaN            NaN  
2           0.935673                 NaN            NaN  
3           0.906433                 NaN            NaN  
4                NaN            0.918129            NaN  
5                NaN            0.935673            NaN  
6                NaN                 NaN       0.865497  
7                NaN                 NaN       0.912281

### Висновки
#### Найкращі показники отримали при застосуванні ансамблю моделей StackingClassifier з базовими моделями LogisticRegression і DecisionTreeClassifier та фінальною моделью SVC.  Score=0.941520	